### PR TITLE
Fixed: $row() undefined issue while inline edit after delete item from database

### LINKED ISSUE #37525: CMS Page && CMS Block edit save error
### Preconditions and environment

- Magento ver. 2.4.5-p1

### Steps to reproduce

1 Go Cms page list
2 Inline Edit Select the last two or three columns , make sure the last column selected
image
3 open another window delete one of the three columns (not the last one column) or delete the column in database
image
4 then save those columns will cause error
image
5 remove the error column
image
6 save another columns (include the last one column),a loading error occurred.still loading 
image
7 the problem same happend in CMS Block with same steps
image

### Expected result

save the normal columns data success and redirct list 

### Actual result

Had been loading without redirect

### Additional information

_No response_

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by aasim110**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#37525

### Manual testing scenarios (*)

1 Go Cms page list
2 Inline Edit Select the last two or three columns, and make sure the last column selected
image
3 open another window delete one of the three columns (not the last column) or delete the column in database
image
4 then save those columns will cause error
image
5 remove the error column
image
6 save another column (including the last one column), a loading error occurred. still loading
image
7 The problem same happened in CMS Block with the same steps
image

### Questions or comments

1. Afrer commit changes continues loading issue has been resolved.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by aasim110** [comment]
@magento run all tests

**Comment by aasim110** [comment]
@magento run all tests

**Comment by aasim110** [comment]
@magento run all tests

**Comment by aasim110** [comment]
@magento please run Functional Tests EE

**Comment by aasim110** [comment]
Hey @ihor-sviziev,
Can someone else please check this?

**Comment by ihor-sviziev** [comment]
@aasim110, please add the steps to reproduce, an actual and expected result sections to issue description.

**Comment by aasim110** [comment]
@ihor-sviziev I've updated the description.

**Comment by aasim110** [comment]
https://github.com/magento run all tests

**Comment by aasim110** [comment]
@magento run all tests

**Comment by aasim110** [comment]
Hi @ihor-sviziev I checked the logs and found that the error related to the Functional test B2B, 
Error: in this file /var/www/html/dev/tests/integration/testsuite/Magento/NegotiableQuote/Cron/ExpireQuoteTest.php:148

So NegotiableQuote and its integration within the B2B suite, which serves as an add-on to our Enterprise offering. I must admit that I am somewhat uncertain about the necessity of functional and integration tests for both the Enterprise Edition (EE) and B2B when submitting a pull request to the Open Source repository. Could we explore the possibility of running solely the Open Source tests instead?

**Comment by aasim110** [comment]
@magento run integration tests

**Comment by aasim110** [comment]
@magento run all tests

**Comment by aasim110** [comment]
@magento run Functional Tests EE

**Comment by aasim110** [comment]
@magento run Functional Tests CE

**Comment by aasim110** [comment]
@magento run Functional Tests EE

**Comment by aasim110** [comment]
@magento run Functional Tests EE

**Comment by aasim110** [comment]
@magento run Functional Tests EE

**Comment by aasim110** [comment]
@magento run all tests

**Comment by aasim110** [comment]
@magento run Functional Tests B2B, Functional Tests CE

**Comment by aasim110** [comment]
@magento run Functional Tests CE

**Comment by aasim110** [comment]
@magento run Functional Tests CE

**Comment by aasim110** [comment]
@magento run Sample Data Tests EE, Sample Data Tests CE, Sample Data Tests B2B

**Comment by manavluhar** [comment]
@magento run Sample Data Tests CE, Sample Data Tests EE, Sample Data Tests B2B

**Comment by manavluhar** [comment]
@magento run Sample Data Tests CE

**Comment by aasim110** [comment]
@magento run all tests

**Comment by aasim110** [comment]
@magento run all tests

**Comment by aasim110** [comment]
@magento run Functional Tests EE, Functional Tests CE, WebAPI Tests

**Comment by aasim110** [comment]
@magento run Functional Tests EE, Functional Tests CE

**Comment by aasim110** [comment]
@magento run all Tests

**Comment by aasim110** [comment]
@magento run Functional Tests CE

**Comment by aasim110** [comment]
@magento run Functional Tests CE

**Comment by aasim110** [comment]
@magento run Functional Tests CE

**Comment by aasim110** [comment]
@magento run Sample Data Tests EE, Sample Data Tests CE, Sample Data Tests B2B

**Comment by aasim110** [comment]
@ihor-sviziev @manavluhar can you please check? why failed to run the builds error comes when running Sample Data Tests.

**Comment by manavluhar** [comment]
@magento run Sample Data Tests CE

**Comment by manavluhar** [comment]
Hello @engcom-Bravo, @engcom-Hotel 
Can someone please see why we are not able to execute **Sample Data Tests** here?
Please help here to proceed further.

**Comment by engcom-Hotel** [comment]
@manavluhar It has been skipped for now. Please ignore it.

**Comment by aasim110** [comment]
@magento run all Tests

**Comment by aasim110** [comment]
@magento run Functional Tests EE

**Comment by aasim110** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by aasim110** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by aasim110** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by aasim110** [comment]
@magento run all Tests

**Comment by aasim110** [comment]
@magento run Integration Tests, Functional Tests CE

**Comment by aasim110** [comment]
@engcom-Hotel @engcom-Bravo I could be wrong, but it seems to me that the failed tests do not refer to edits from the PR.

**Comment by aasim110** [comment]
@magento run Functional Tests CE

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @aasim110,

Thanks for the contribution and collaboration!

Code changes seems good to me and failed tests seem flaky to me. But can we cover these changes with some automated tests to ensure the fix.

Thanks

**Comment by engcom-Charlie** [comment]
Hi @aasim110,

Thank you for your contribution!

In order to proceed further on this PR, I have tried to check the PR implementation and found that the mentioned issue is not yet resolved.

1. Visited CMS page list
2. Selected the last 3 rows, clicked on Actions >> Edit section
image

3. Duplicate the window and tried to Delete the second last record.
image

4. Click on Delete
image

5. Came back on the first window and tried to Save. At this point, on PR branch we can still see the same error mentioned in PR description. 
image


@aasim110, can you please have a look into this and let me know if we are missing anything here in order to check this.

Thank you!

**Comment by aasim110** [comment]
@engcom-Charlie the issue is **infinite loading** after deleting that row, as per your screenshot you have deleted page id = 7, and after deleting you are trying to save the same page and an error comes which is correct. but previously when you are trying to save the infinite loading comes and that is what I have resolved in this PR.  
Please check all steps given by me again to replicate the issue before this PR: https://github.com/magento/magento2/pull/37560#issue-1732700168

**Comment by engcom-Charlie** [comment]
Thank you for the quick reply @aasim110 

I have checked the mentioned issue with PR changes. After following steps mentioned in [description](https://github.com/magento/magento2/pull/37560#issue-1732700168) the PR is working fine but there is one more test scenario which is still showing the error.

Scenario: If we delete 2 rows excluding the last row and try to save as mentioned in description, the buffering issue still persist there. Can you please have a look into it

Thank you!

**Comment by aasim110** [comment]
@engcom-Charlie kindly use my https://github.com/magento/magento2/pull/37560/commits/85e2f141d791a5fecca89932fa6a0d71d1e6545a commit to resolve multiple save issue

Thanks!

**Comment by engcom-Charlie** [comment]
It seems that there are some changes done in implementation of this PR, hence moving it to Review again.

Thank you!

**Comment by engcom-Hotel** [comment]
Hello @aasim110,

Thanks for the reply!

As we can see there are changes happens in this PR, but still facing issues with another scenario mentioned here https://github.com/magento/magento2/pull/37560#issuecomment-1620288159.

I can understand that you have fixed that scenario in this commit https://github.com/magento/magento2/commit/85e2f141d791a5fecca89932fa6a0d71d1e6545a, but we are unable to see this commit in this PR. 

So we request you to please add that commit to this PR so that we will proceed with this as per the process. Meanwhile marking this PR as `On Hold.`

Thanks

**Comment by aasim110** [comment]
@engcom-Hotel Could you please check now?

File: app/code/Magento/Ui/view/base/web/js/grid/editing/editor-view.js [hunk 1]
```diff
@@ -28,12 +28,18 @@ define([
...
                 '<!-- ko with: _editor -->' +
-                    '<!-- ko if: isActive($row()._rowIndex, true) -->' +
-                        '<!-- ko with: getRecord($row()._rowIndex, true) -->' +
-                            '<!-- ko template: rowTmpl --><!-- /ko -->' +
-                        '<!-- /ko -->' +
-                        '<!-- ko if: isSingleEditing && singleEditingButtons -->' +
-                            '<!-- ko template: rowButtonsTmpl --><!-- /ko -->' +
+                    '<!-- ko if: typeof $row() !== "undefined" -->' +
+                        '<!-- ko if: isActive($row()._rowIndex, true) -->' +
+                            '<!-- ko if: typeof $row() !== "undefined" -->' +
```
Comment (engcom-Hotel) [inline app/code/Magento/Ui/view/base/web/js/grid/editing/editor-view.js:33]: I think we can remove this condition as we are already checking the same in line number 31.

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @aasim110,

Thanks for the collaboration!

Kindly review the comment below.

Thanks

File: app/code/Magento/Ui/view/base/web/js/grid/editing/editor-view.js (hunk 1 above)
Comment (aasim110) [reply app/code/Magento/Ui/view/base/web/js/grid/editing/editor-view.js:33]: @engcom-Hotel I cannot remove the code that checks for undefined rows. This code is necessary to check each column for undefined values, as the $row() method will return undefined if no rows are found. If we remove this code, the $row() undefined issue will come back.

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Unit Tests ,Static Tests

**Comment by engcom-Lima** [comment]
@magento run Functional Tests EE, Static Tests

**Comment by engcom-Lima** [comment]
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`** and **`PHP 8.1`**

**Manual testing scenario:**
1 Go Cms page list
2 Inline Edit Select the last two or three columns and make sure the last column selected
3 open another window delete one of the three columns (not the last column) or delete the column in the database
4 then save those columns will cause an error.
5 remove the error column.
6 Save another column (including the last column), a loading error occurred. still loading
7 The problem same happened in **CMS Block** with the same steps.

**Before:** :heavy_multiplication_x: Page loading issue.
Screenshot 2023-07-17 at 3 09 07 PM


**After:** :heavy_check_mark:  
Screenshot 2023-07-17 at 3 42 41 PM
Screenshot 2023-07-18 at 2 45 46 PM

**Additionally tested:**
- Verified the grid functionality of Pages and Blocks.

Builds are failed. Hence, moving this PR to **Extended Testing.**

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE, Static Tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests EE

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Static Tests

**Comment by engcom-Echo** [comment]
@magento run Static Tests

**Comment by engcom-Echo** [comment]
@magento run Unit Tests,Integration Tests,Functional Tests EE,Functional Tests CE,Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Static Tests

**Comment by engcom-Echo** [comment]
@magento run Static Tests

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests EE ,Functional Tests CE, Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Functional Tests EE,Functional Tests CE,Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magneto run Functional Tests CE,Functional Tests EE,Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magneto run Functional Tests CE,Functional Tests EE,Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magneto run Functional Tests CE,Functional Tests EE,Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B,Functional Tests CE,Functional Tests EE

**Comment by engcom-Echo** [comment]
Failing functional test are not related to this PR.
Hence Moving to `merge in progress.`

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B,Functional Tests EE,Integration Tests,Static Tests,Unit Tests ,WebAPI Tests

**Comment by engcom-Echo** [comment]
Functional Tests EE failure are different on last two run on same commit. Other build failure are not related to PR changes.
Functional Tests EE
Run1
37560-ee

Run2
Screenshot 2023-08-18 at 12 33 16 PM

**Pull request merged**